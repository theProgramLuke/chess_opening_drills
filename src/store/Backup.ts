import _ from "lodash";
import fs from "graceful-fs";
import path from "path";

export const AgeSeparator = "-";

export class Backup {
  filePath: string;
  age: number;

  constructor(filePath: string) {
    this.filePath = filePath;
    this.age = _.parseInt(this.filePath.split(AgeSeparator)[1]);
  }

  async delete(): Promise<void> {
    await fs.unlink(this.filePath, _.noop);
  }

  async save(content: string): Promise<void> {
    await fs.mkdir(
      path.dirname(this.filePath),
      { recursive: true },
      async () => {
        await fs.writeFile(this.filePath, content, _.noop);
      }
    );
  }
}
