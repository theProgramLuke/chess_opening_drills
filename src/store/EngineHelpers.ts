import _ from "lodash";
import { Engine } from "node-uci";

interface SourceEngineOption {
  type: "spin" | "check" | "combo" | "string" | "button";
  default: number | boolean | string;
  value: number | boolean | string;
  min?: number;
  max?: number;
  options?: string[];
}

export interface EngineOption extends SourceEngineOption {
  name: string;
}

export interface EngineMetadata {
  name: string;
  filePath: string;
  options: EngineOption[];
}

interface MetadataEngine extends Engine {
  options?: Map<string, SourceEngineOption>;
  id?: { name: string };
}

export async function GetMetadataFromEngine(
  enginePath: string
): Promise<EngineMetadata | undefined> {
  if (enginePath) {
    // Launch the engine with a dummy position to get the available options.
    const engine: MetadataEngine = new Engine(enginePath);

    await engine.init();
    await engine.position("7K/8/8/8/8/8/8/7k");
    await engine.go({ depth: 1 });

    if (engine.options && engine.id) {
      const options: EngineOption[] = [];
      engine.options.forEach((option, key) => {
        if (option.type !== "button") {
          option.value = option.default;
          options.push({
            name: key,
            ...option
          });
        }
      });

      return {
        name: engine.id.name,
        filePath: enginePath,
        options
      };
    }
  }

  return undefined;
}

export interface EngineOutput {
  evaluation: number;
  depth: number;
  variation: string[];
  id: number;
}

export function ProcessAnalysis(engineData: any): EngineOutput | undefined {
  if (engineData.depth && engineData.score && engineData.pv) {
    return {
      evaluation: engineData.score.value / 100,
      depth: engineData.depth,
      variation: engineData.pv.split(" "),
      id: engineData.multipv
    };
  }

  return undefined;
}
