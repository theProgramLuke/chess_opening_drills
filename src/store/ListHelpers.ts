import _ from "lodash";

function IsPrefix<T>(list: T[], potentialPrefix: T[]): boolean {
  if (potentialPrefix.length > list.length) {
    return false;
  }

  let foundMismatch = false;
  _.forEach(potentialPrefix, (entry, index) => {
    if (!foundMismatch && entry !== list[index]) {
      foundMismatch = true;
    }
  });

  return !foundMismatch;
}

export function filterPrefixLists<T>(lists: T[][]): T[][] {
  return _.filter(lists, potentialPrefix => {
    let foundAsPrefix = false;
    _.forEach(lists, moveList => {
      if (moveList !== potentialPrefix) {
        foundAsPrefix = foundAsPrefix || IsPrefix(moveList, potentialPrefix);
      }
    });
    return !foundAsPrefix;
  });
}
