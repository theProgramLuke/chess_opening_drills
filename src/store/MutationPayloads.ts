import { Repertoire } from "@/store/repertoire/Repertoire";
import { TrainingEvent } from "@/store/repertoire/RepetitionTraining";
import { TagTree } from "@/store/repertoire/TagTree";

export interface AddTrainingEventPayload {
  repertoire: Repertoire;
  fen: string;
  san: string;
  event: TrainingEvent;
}

export interface AddRepertoireMovePayload {
  repertoire: Repertoire;
  fen: string;
  san: string;
}

export interface SetColorPayload {
  colorToSet:
    | "primary"
    | "secondary"
    | "accent"
    | "error"
    | "warning"
    | "info"
    | "success";
  color: string;
}

export interface AddRepertoireTagPayload {
  repertoire: Repertoire;
  parent: TagTree;
  name: string;
  fen: string;
}

export interface RemoveRepertoireMovePayload {
  repertoire: Repertoire;
  fen: string;
  san: string;
}

export interface AddMovesFromPgnPayload {
  repertoire: Repertoire;
  pgn: string;
}

export interface RemoveRepertoireTagPayload {
  repertoire: Repertoire;
  id: string;
}
