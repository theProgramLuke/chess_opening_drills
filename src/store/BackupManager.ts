import fs from "graceful-fs";
import path from "path";
import _ from "lodash";

import { Backup } from "@/store/Backup";

export const day = 24 * 60 * 60 * 1000;
export const month = day * 31;
export const year = day * 365;

const lastBackupAge = (backups: Backup[]): number => {
  return _.min(_.map(backups, backup => backup.age)) || 0;
};

const trimBackups = (backups: Backup[], limit: number): Backup[] => {
  const sorted = _.sortBy(backups, backup => 0 - backup.age);

  while (sorted.length > limit) {
    const toDelete = sorted.pop();
    if (toDelete) {
      toDelete.delete();
    }
  }

  return sorted;
};

export class BackupManager {
  backupFolder: string;
  dailyBackups: Backup[];
  monthlyBackups: Backup[];
  yearlyBackups: Backup[];
  dailyLimit: number;
  monthlyLimit: number;
  yearlyLimit: number;
  createBackup: (filePath: string) => Backup;

  constructor(
    backupFolder: string,
    dailyLimit: number,
    monthlyLimit: number,
    yearlyLimit: number,
    listDirectory = (directory: string): string[] => fs.readdirSync(directory),
    createBackup = (filePath: string) => new Backup(filePath)
  ) {
    this.backupFolder = backupFolder;
    this.createBackup = createBackup;
    this.dailyBackups = this.discoverBackups(
      "daily",
      listDirectory,
      createBackup
    );
    this.monthlyBackups = this.discoverBackups(
      "monthly",
      listDirectory,
      createBackup
    );
    this.yearlyBackups = this.discoverBackups(
      "yearly",
      listDirectory,
      createBackup
    );
    this.dailyLimit = dailyLimit;
    this.monthlyLimit = monthlyLimit;
    this.yearlyLimit = yearlyLimit;
  }

  private discoverBackups(
    frequency: string,
    listDirectory: (directory: string) => string[],
    createBackup: (filePath: string) => Backup
  ): Backup[] {
    try {
      const files = listDirectory(path.join(this.backupFolder, frequency));
      return _.map(files, createBackup);
    } catch {
      return [];
    }
  }

  SaveBackup(content: string, now = _.now): void {
    const fileName = `settings-${now()}.json`;
    let backup: Backup;

    if (this.dailyLimit && now() - lastBackupAge(this.dailyBackups) >= day) {
      backup = this.createBackup(
        path.join(this.backupFolder, "daily", fileName)
      );
      this.dailyBackups.push(backup);
      backup.save(content);
    }
    this.dailyBackups = trimBackups(this.dailyBackups, this.dailyLimit);

    if (
      this.monthlyLimit &&
      now() - lastBackupAge(this.monthlyBackups) >= month
    ) {
      backup = this.createBackup(
        path.join(this.backupFolder, "monthly", fileName)
      );
      this.monthlyBackups.push(backup);
      backup.save(content);
    }
    this.monthlyBackups = trimBackups(this.monthlyBackups, this.monthlyLimit);

    if (this.yearlyLimit && now() - lastBackupAge(this.yearlyBackups) >= year) {
      backup = this.createBackup(
        path.join(this.backupFolder, "yearly", fileName)
      );
      this.yearlyBackups.push(backup);
      backup.save(content);
    }
    this.yearlyBackups = trimBackups(this.yearlyBackups, this.yearlyLimit);
  }
}
