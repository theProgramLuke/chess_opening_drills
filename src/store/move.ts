import { RepertoirePosition } from "@/store/repertoirePosition";

export class Move {
  san: string;
  position: RepertoirePosition;

  constructor(san: string, position: RepertoirePosition) {
    this.san = san;
    this.position = position;
  }
}
