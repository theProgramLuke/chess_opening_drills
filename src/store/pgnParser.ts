import { parse } from "@/store/pgnGrammar";

export interface PgnMove {
  moveNumber?: number;
  move?: PgnMove;
  nags?: string[];
  ravs?: string[];
  comments?: string;
}

export interface PgnHeader {
  name: string;
  value: string;
}

export interface PgnRav {
  moves: PgnMove[];
  result: string;
}

export interface PgnGame {
  commentsAboveHeader: string;
  headers: PgnHeader[];
  comments: string;
  moves: PgnMove[];
  result: string;
}

export function parsePgn(pgnText: string): PgnGame | string {
  try {
    return parse(pgnText) as PgnGame;
  } catch (error) {
    return error.message;
  }
}
