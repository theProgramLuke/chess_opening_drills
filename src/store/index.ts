import Vue from "vue";
import Vuex from "vuex";

import { RepertoirePosition } from "@/store/repertoirePosition";
import { Move } from "./move";
import { GetPersistantStorage } from "./storage";
import { Repertoire } from "./repertoire";
import { RepertoireTag } from "./repertoireTag";
import { Side } from "./side";
import { TrainingEvent } from "./TrainingEvent";
import { PgnGame } from "./pgnParser";

Vue.use(Vuex);

const storage = GetPersistantStorage();

export default new Vuex.Store({
  state: {
    darkMode: storage.get("darkMode"),
    primary: storage.get("primary"),
    secondary: storage.get("secondary"),
    accent: storage.get("accent"),
    error: storage.get("error"),
    warning: storage.get("warning"),
    info: storage.get("info"),
    success: storage.get("success"),
    boardTheme: storage.get("boardTheme"),
    pieceTheme: storage.get("pieceTheme"),
    whiteRepertoire: Repertoire.FromSaved(storage.get("whiteRepertoire")),
    blackRepertoire: Repertoire.FromSaved(storage.get("blackRepertoire"))
  },

  mutations: {
    setDarkMode: (state, darkMode): void => {
      state.darkMode = darkMode;
      storage.set("darkMode", darkMode);
    },

    setBoardTheme: (state, boardTheme): void => {
      state.boardTheme = boardTheme;
      storage.set("boardTheme", boardTheme);
    },

    setPieceTheme: (state, pieceTheme): void => {
      state.pieceTheme = pieceTheme;
      storage.set("pieceTheme", pieceTheme);
    },

    setColor: function(
      state,
      payload: {
        colorToSet:
          | "primary"
          | "secondary"
          | "accent"
          | "error"
          | "warning"
          | "info"
          | "success";
        color: string;
      }
    ): void {
      state[payload.colorToSet] = payload.color;
      storage.set(payload.colorToSet, payload.color);
    },

    addRepertoirePosition: (
      state,
      payload: { parent: RepertoirePosition; newMove: Move }
    ): void => {
      if (payload.parent && payload.newMove) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddMove(payload.parent, payload.newMove);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addRepertoireTag: (
      state,
      payload: { parent: RepertoireTag; tag: RepertoireTag }
    ): void => {
      if (payload.parent && payload.tag) {
        const repertoire =
          payload.parent.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.parent.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.parent.AddChild(payload.tag);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    addTrainingEvent: (
      state,
      payload: { position: RepertoirePosition; event: TrainingEvent }
    ): void => {
      if (payload.position && payload.event) {
        const repertoire =
          payload.position.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.position.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        payload.position.AddTrainingEvent(payload.event);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    removeRepertoireTag: (state, tag: RepertoireTag): void => {
      const repertoire =
        tag.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        tag.forSide === Side.White ? "whiteRepertoire" : "blackRepertoire";

      repertoire.RemoveRepertoireTag(tag);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    removeRepertoireMove: (state, move: Move): void => {
      const repertoire =
        move.position.forSide === Side.White
          ? state.whiteRepertoire
          : state.blackRepertoire;
      const repertoireKey =
        move.position.forSide === Side.White
          ? "whiteRepertoire"
          : "blackRepertoire";

      repertoire.RemoveMove(move);

      storage.set(repertoireKey, repertoire.AsSaved());
    },

    addPositionsFromGame(
      state,
      payload: { forSide: Side; game: PgnGame }
    ): void {
      if (payload.game) {
        const repertoire =
          payload.forSide === Side.White
            ? state.whiteRepertoire
            : state.blackRepertoire;
        const repertoireKey =
          payload.forSide === Side.White
            ? "whiteRepertoire"
            : "blackRepertoire";

        repertoire.AddFromGame(payload.game);

        storage.set(repertoireKey, repertoire.AsSaved());
      }
    },

    clearStorage(): void {
      storage.clear();
    }
  },

  actions: {},

  modules: {},

  strict: process.env.NODE_ENV !== "production"
});
