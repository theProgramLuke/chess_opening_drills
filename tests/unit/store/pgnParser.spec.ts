import { parsePgn } from "@/store/pgnParser";

describe("parse", () => {
  it("should get the PGN of a game", () => {
    const parsed = parsePgn(
      '[White "me"]\n[Black "you"]\n1. e4 e5 2. Nf3 Nc6 3. Bc4 Bc5 (3. ...Nf6 {is the two knights}) 4. b4 Bxb4 5. c3 Ba5 6. d4 exd4 7. O-O Nge7 $1 *'
    );

    expect(parsed).toMatchSnapshot();
  });

  it("should fail for invalid PGN", () => {
    const parse = () => parsePgn("invalid");

    expect(parse).toThrowError(
      'Expected " ", "*", ".", "0-1", "1-0", "1/2-1/2", ";", "O-O", "O-O-O", "[", "\\n", "\\r", "\\t", "\\x0C", "{", [1-9], [N,K,Q,R,B], or [a-h] but "i" found.'
    );
  });
});
