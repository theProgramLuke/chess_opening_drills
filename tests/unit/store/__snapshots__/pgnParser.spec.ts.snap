// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should get the PGN of a game 1`] = `
Array [
  Object {
    "comments": null,
    "commentsAboveHeader": null,
    "headers": Array [
      Object {
        "name": "White",
        "value": "me",
      },
      Object {
        "name": "Black",
        "value": "you",
      },
    ],
    "moves": Array [
      Object {
        "move": "e4",
        "moveNumber": 1,
      },
      Object {
        "move": "e5",
      },
      Object {
        "move": "Nf3",
        "moveNumber": 2,
      },
      Object {
        "move": "Nc6",
      },
      Object {
        "move": "Bc4",
        "moveNumber": 3,
      },
      Object {
        "move": "Bc5",
        "ravs": Array [
          Object {
            "moves": Array [
              Object {
                "comments": "is the two knights",
                "move": "...Nf6",
                "moveNumber": 3,
              },
            ],
            "result": null,
          },
        ],
      },
      Object {
        "move": "b4",
        "moveNumber": 4,
      },
      Object {
        "move": "Bxb4",
      },
      Object {
        "move": "c3",
        "moveNumber": 5,
      },
      Object {
        "move": "Ba5",
      },
      Object {
        "move": "d4",
        "moveNumber": 6,
      },
      Object {
        "move": "exd4",
      },
      Object {
        "move": "O-O",
        "moveNumber": 7,
      },
      Object {
        "move": "Nge7",
        "nags": Array [
          "$1",
        ],
      },
    ],
    "result": "*",
  },
]
`;
